internal/operations/manager.go

```diff
@@ -0,0 +1,148 @@
+// Copyright © 2022 Kaleido, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operations
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hyperledger/firefly/internal/i18n"
+	"github.com/hyperledger/firefly/internal/txcommon"
+	"github.com/hyperledger/firefly/pkg/database"
+	"github.com/hyperledger/firefly/pkg/fftypes"
+	"github.com/hyperledger/firefly/pkg/tokens"
+)
+
+type Manager interface {
+	StartOperation(ctx context.Context, op *fftypes.Operation) error
+}
+
+type operationsManager struct {
+	ctx      context.Context
+	database database.Plugin
+	txHelper txcommon.Helper
+	tokens   map[string]tokens.Plugin
+}
+
+func NewOperationsManager(ctx context.Context, di database.Plugin, ti map[string]tokens.Plugin) (Manager, error) {
+	if di == nil || ti == nil {
+		return nil, i18n.NewError(ctx, i18n.MsgInitializationNilDepError)
+	}
+	om := &operationsManager{
+		ctx:      ctx,
+		database: di,
+		txHelper: txcommon.NewTransactionHelper(di),
+		tokens:   ti,
+	}
+	return om, nil
+}
+
+func (om *operationsManager) StartOperation(ctx context.Context, op *fftypes.Operation) error {
```

Comment:
Agree we can't afford to have repeated reads on the critical path, if the calling code already has all this in hand.

I also agree that a suite of E2E tests would be the most durable way of testing these code paths, but clearly that comes with an initial cost of introducing some systematic way of introducing failures at the right phase.

I wonder if a cost-effective compromise would be a REST API call for `GET` `/{ns}/transactions/{txid}operations/{opid}/inputs` that the E2E test uses to confirm the JSON matches what's expected. This wouldn't need any guards around it.

Even typing this, I understand it's complicated - because this would mean only immutable parts of data objects are used to construct the inputs. e.g. if the operation succeeds, and the state changes, it's still possible to generate the _original_ inputs of the operation. This feels "right" in a blockchain centric system, but I understand in practice there could still be complexities.

A rambling thought on the code structure, that I don't feel strongly about but might be useful in your ideas:
- An defined input structure for each type of operation, containing the source types that are used to aggregate together the input
- The assembly code that does the mapping is shared between submit and retry

Reply:
After I wrote this, I realized that today we have a DB persisted `inputs` field on the operation itself.

I suddenly feel like I'm missing something obvious. Why are the inputs not always available on re-submission, simply because at the submission time they were stored in the DB?

Reply:
Makes sense. Particularly for something like an IPFS batch, where we're currently trying to avoid writing the batch to the DB once... and this might end up with us doing it twice if we're not careful.

Reply:
Had a bit of a further chat with @peterbroadhurst 1:1, so for the record:
Noted that there definitely are two phases here ("loading/preparing" the needed data from the operation input and any related objects, and then actually "running" the operation). The loading part can be skipped on the initial trigger if the caller already has all the needed information pre-built. Already agreed on this part.

We further identified that the "loading" and "preparing" could be split apart, and we should define some standard format for the "prepared" operation. It should be a serializable object that expresses something at the granularity of "this wraps an HTTP GET to /xyz with params A, B, C". Exact formatting may be at the discretion of the plugin, but each operation should have a "prepare" and a "run" action. Defining an intermediate format here will be useful for debugging, E2E testing, and eventually for more robust retry flows such as "prepare-modify-run".